File: neetcode all and misc questions/Math/plus_one.py
"""
Revisited on 5/5/2025
Problem found on intuit's list: 
https://github.com/liquidslr/leetcode-company-wise-problems/blob/main/Intuit/2.%20Three%20Months.csv
"""
class Solution:
    def plusOne(self, digits: List[int]) -> List[int]:
        """
        you can convert the numbers to strings first, perform string concatentation,
        and then check if there's a carryover after adding one
        and then convert back to string, split into an array

        can also iterate from the back of the array and add the numbers without needing
        to perform string concatenation

        [1,2,3]
        3 * (10*0)
        2 * (10^1)
        1 * (10^2)
        -----
        3 + 20 + 100 = 123
        O(2N) Time
        O(1) Space

        Can also keep as an array, and then perform the plus one operation directly
        on the array numbers starting from the back
        O(N) Time
        O(1) Space
        """

        # num = 0
        # j = 0
        # for i in range(len(digits)-1,-1,-1):
        #     num += (digits[i] * (10**(j)))
        #     j+=1
        # num += 1
        # res = []
        # while (num > 0):
        #     digit = num % 10
        #     res = [digit] + res
        #     num = num//10
        # return res

        carryover = 0
        res = [0] * len(digits)
        for i in range(len(digits)-1,-1,-1):
            curSum = digits[i] + 1 if i == len(digits) - 1 else carryover + digits[i]
            if curSum >= 10:
                carryover = 1
                res[i] = 0
            else:
                carryover = 0
                res[i] = curSum
        # in the case where there's still a carryover after iterating through all digits,
        # include the last digit
        if (carryover == 1):
            res = [carryover] + res
        return res
            
"""
https://leetcode.com/problems/plus-one/
simplified version of add strings
"""
class Solution:
    def plusOne(self, digits: List[int]) -> List[int]:
        carryover = 0
        output = []
        for i in range(len(digits)-1,-1,-1):
            s = digits[i] + carryover + 1 if i == len(digits)-1 else digits[i] + carryover
            if s >= 10:
                carryover = 1
                output.append(s - 10)
            else:
                output.append(s)
                carryover = 0
        if carryover != 0:
            output.append(carryover)
        output.reverse()
        return output
        

File: node_js_ds_and_algos_implementations/math and geometry/plus_one.js
class Solution {
    /**
     * @param {number[]} digits
     * @return {number[]}
     */
    plusOne(digits) {
        // revisited 7/16/2026
        let carryover = 0
        // if the last digit is 9, there will be carryover
        if (digits[digits.length-1] === 9){
            let i = digits.length-1
            // here, we assume that we add "plus one", so since we're starting from 
            // the last element, we can just assume a carryover of 1 to simulate that
            let carryover = 1
            while (i >= 0){
                let res = digits[i] + carryover
                if (res >= 10){
                    carryover = 1
                    digits[i] = res - 10
                    --i
                }
                else {
                    digits[i] = res
                    carryover = 0
                    break
                }
            }
            if (carryover === 1){
                digits.unshift(1)
            }
        }
        else {
            digits[digits.length-1] += 1
        }
        return digits
    }
}

/**
 * @param {number[]} digits
 * @return {number[]}
 */
var plusOne = function(digits) {
    let i = digits.length-1
    // initially add one to the last digit
    let carryover = 0
    if (digits[i] + 1 >= 10){
        carryover = 1
        digits[i] = 0
    }
    else {
        carryover = 0
        digits[i] = digits[i] + 1
    }
    --i
    // if we're not at the front of the array, and there's carryover,
    // continuing adding the carryover until it's no longer 1
    // for example: 1, 9, 9
    // 2, 0, 0,
    // where the 1 is carried over after the initial addition of 1 to 9
    while (i >= 0 && carryover === 1){
        if (digits[i] + carryover >= 10){
            carryover = 1
            digits[i] = 0
        }
        else {        
            digits[i] = digits[i] + carryover
            carryover = 0
        }
        --i
    }
    // if there's still carryover, but we've run out of digits, we just add the carryover
    // to the beginning of the array
    // i.e [9], we'd add 1 to get 10, but there's a carryover of 1 and we're at the front of the array,
    // so this would be [1, 0]
    if (carryover === 1){
        digits.unshift(carryover)
    }
    return digits
};